[quant][graphmode][fx] Make maxpool and flatten produce the reference pattern

jerryzh168 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #63828
* #63627
* __->__ #63501
* #62861
* #62698

Summary:
Currently some of the ops are considered as working with both float and quantized input,
so we may have things like "quant - some_op - dequant" this might not work well with the backend,
we may consider change everything to produce "quant - dequant - some_op - quant - dequant" instead
in the future, this PR fixes it for maxpool and flatten only to unblock resnet benchmarking on TensorRT

Test Plan:
python test/test_quantization.py TestQuantizeFxOps

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D30402788](https://our.internmc.facebook.com/intern/diff/D30402788)

jerryzh168 committed:
[quant][graphmode][fx] Make maxpool and flatten produce the reference pattern

Summary:
Currently some of the ops are considered as working with both float and quantized input,
so we may have things like "quant - some_op - dequant" this might not work well with the backend,
we may consider change everything to produce "quant - dequant - some_op - quant - dequant" instead
in the future, this PR fixes it for maxpool and flatten only to unblock resnet benchmarking on TensorRT

Test Plan:
python test/test_quantization.py TestQuantizeFxOps

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 commented:
@jerryzh168 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30402788).

jerryzh168 committed:
Update on "[quant][graphmode][fx] Make maxpool and flatten produce the reference pattern"


Summary:
Currently some of the ops are considered as working with both float and quantized input,
so we may have things like "quant - some_op - dequant" this might not work well with the backend,
we may consider change everything to produce "quant - dequant - some_op - quant - dequant" instead
in the future, this PR fixes it for maxpool and flatten only to unblock resnet benchmarking on TensorRT

Test Plan:
python test/test_quantization.py TestQuantizeFxOps

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D30402788](https://our.internmc.facebook.com/intern/diff/D30402788)

[ghstack-poisoned]

jerryzh168 commented:
@jerryzh168 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30402788).

raghuramank100 commented on torch/quantization/fx/quantization_patterns.py:
```diff
@@ -1456,12 +1459,16 @@ def convert(self,
 @register_quant_pattern(torch.nn.functional.hardtanh)
 @register_quant_pattern(torch.nn.functional.hardtanh_)
 @register_quant_pattern(torch.nn.functional.interpolate)
+@register_quant_pattern(torch.nn.functional.max_pool1d)
```
For my understanding, the CopyNodeQuantizeHandler is used for many patterns. How is it that only maxpool1d/2d/3d produces a pattern with a dequant->maxpool->quant? What about other operators like clamp?

jerryzh168 replied:
Yeah I think we should probably remove GeneralTensorValueQuantizeHandler and move everything here, it looks like some backend may not work when we have "quant - some_general_op - dequant", TensorRT included. I think we should just produce "dequant - float_op - quant" pattern for all these ops.
Planning to change that in a separate PR, current goal is to test TensorRT integration.

raghuramank100 commented on test/quantization/fx/test_quantize_fx.py:
```diff
@@ -4171,8 +4171,8 @@ def forward(self, x):
         count_check = {
-            ns.call_function(torch.quantize_per_tensor) : 1,
-            ns.call_method('dequantize') : 1
+            ns.call_function(torch.quantize_per_tensor) : 2,
```
Will a more modular test help? Currently it is not clear what is exactly being tested.

jerryzh168 replied:
oh, this is testing that we do not have extra quant/dequant for all the ops listed in the forward function, these ops are what we call "CopyNode" in the quantize handler, this test might not be very useful in the future

raghuramank100 reviewed (approved):
Looks good, but I think I am missing overall context.

jerryzh168 committed:
Update on "[quant][graphmode][fx] Make maxpool and flatten produce the reference pattern"


Summary:
Currently some of the ops are considered as working with both float and quantized input,
so we may have things like "quant - some_op - dequant" this might not work well with the backend,
we may consider change everything to produce "quant - dequant - some_op - quant - dequant" instead
in the future, this PR fixes it for maxpool and flatten only to unblock resnet benchmarking on TensorRT

Test Plan:
python test/test_quantization.py TestQuantizeFxOps

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D30402788](https://our.internmc.facebook.com/intern/diff/D30402788)

[ghstack-poisoned]

jerryzh168 commented:
@jerryzh168 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30402788).

jerryzh168 committed:
Update on "[quant][graphmode][fx] Make maxpool and flatten produce the reference pattern"


Summary:
Currently some of the ops are considered as working with both float and quantized input,
so we may have things like "quant - some_op - dequant" this might not work well with the backend,
we may consider change everything to produce "quant - dequant - some_op - quant - dequant" instead
in the future, this PR fixes it for maxpool and flatten only to unblock resnet benchmarking on TensorRT

Test Plan:
python test/test_quantization.py TestQuantizeFxOps

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D30402788](https://our.internmc.facebook.com/intern/diff/D30402788)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode][fx] Make maxpool and flatten produce the reference pattern"


Summary:
Currently some of the ops are considered as working with both float and quantized input,
so we may have things like "quant - some_op - dequant" this might not work well with the backend,
we may consider change everything to produce "quant - dequant - some_op - quant - dequant" instead
in the future, this PR fixes it for maxpool and flatten only to unblock resnet benchmarking on TensorRT

Test Plan:
python test/test_quantization.py TestQuantizeFxOps

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D30402788](https://our.internmc.facebook.com/intern/diff/D30402788)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode][fx] Make maxpool and flatten produce the reference pattern"


Summary:
Currently some of the ops are considered as working with both float and quantized input,
so we may have things like "quant - some_op - dequant" this might not work well with the backend,
we may consider change everything to produce "quant - dequant - some_op - quant - dequant" instead
in the future, this PR fixes it for maxpool and flatten only to unblock resnet benchmarking on TensorRT

Test Plan:
python test/test_quantization.py TestQuantizeFxOps

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D30402788](https://our.internmc.facebook.com/intern/diff/D30402788)

[ghstack-poisoned]

jerryzh168 commented:
@jerryzh168 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30402788).

jerryzh168 committed:
Update on "[quant][graphmode][fx] Make maxpool and flatten produce the reference pattern"


Summary:
Currently some of the ops are considered as working with both float and quantized input,
so we may have things like "quant - some_op - dequant" this might not work well with the backend,
we may consider change everything to produce "quant - dequant - some_op - quant - dequant" instead
in the future, this PR fixes it for maxpool and flatten only to unblock resnet benchmarking on TensorRT

Test Plan:
python test/test_quantization.py TestQuantizeFxOps

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D30402788](https://our.internmc.facebook.com/intern/diff/D30402788)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode][fx] Make maxpool and flatten produce the reference pattern"


Summary:
Currently some of the ops are considered as working with both float and quantized input,
so we may have things like "quant - some_op - dequant" this might not work well with the backend,
we may consider change everything to produce "quant - dequant - some_op - quant - dequant" instead
in the future, this PR fixes it for maxpool and flatten only to unblock resnet benchmarking on TensorRT

Test Plan:
python test/test_quantization.py TestQuantizeFxOps

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D30402788](https://our.internmc.facebook.com/intern/diff/D30402788)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode][fx] Make maxpool and flatten produce the reference pattern"


Summary:
Currently some of the ops are considered as working with both float and quantized input,
so we may have things like "quant - some_op - dequant" this might not work well with the backend,
we may consider change everything to produce "quant - dequant - some_op - quant - dequant" instead
in the future, this PR fixes it for maxpool and flatten only to unblock resnet benchmarking on TensorRT

Test Plan:
python test/test_quantization.py TestQuantizeFxOps

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D30402788](https://our.internmc.facebook.com/intern/diff/D30402788)

[ghstack-poisoned]
